### PR TITLE
[DirectX] fix illegal behavior flag in module flags.

### LINKED ISSUE #90773: [DirectX] remove module flag and function attribute not supported by DXIL
The only module flags DXIL used are "Dwarf Version" and "Debug Info Version".

And the only string function attributes allowed in DXIL are "waveops-include-helper-lanes" and "fp32-denorm-mode".

Other string function attributes and module flags should be removed before emit DXIL.

### PR DESCRIPTION
**Pull request by python3kgae**
For DXIL which is based on llvm 3.7, max supported behavior flag for module flags is 6.

The commit will check all module flags, for behavior flag > 6, change it to 2 (Warning).
    
This is to fix the behavior flag part for #96912.

### CODE REVIEW

File: llvm/test/CodeGen/DirectX/clean-module-flags2.ll [hunk 1]
```diff
@@ -0,0 +1,17 @@
+; RUN: opt -S -dxil-prepare < %s | FileCheck %s
```
Comment (damyanp) [inline llvm/test/CodeGen/DirectX/clean-module-flags2.ll:1]: It looks like `clean-module-flags.ll` and `clean-module-flags2.ll` are testing the same thing?  Or am I missing something?

File: llvm/test/CodeGen/DirectX/clean-module-flags.ll [hunk 2]
```diff
@@ -0,0 +1,20 @@
+; RUN: opt -S -dxil-prepare < %s | FileCheck %s
+
+; Make sure behavior flag > 6 is fixed.
+; CHECK-NOT: !{i32 7, !"frame-pointer", i32 2}
```
Comment (damyanp) [inline llvm/test/CodeGen/DirectX/clean-module-flags.ll:4]: Interested in why:

* no CHECK line for wchar_size?
* CHECK-NOT is used for frame-pointer, but CHECK is used for Dwarf Version

Comment (python3kgae) [reply llvm/test/CodeGen/DirectX/clean-module-flags.ll:4]: For DXIL, it is legal to not have wchar_size and frame-pointer.
But remove Dwarf Version will lost information for DXIL.

File: llvm/test/CodeGen/DirectX/clean-module-flags2.ll (hunk 1 above)
Comment (python3kgae) [reply llvm/test/CodeGen/DirectX/clean-module-flags2.ll:1]: clean-module-flags.ll has Dwarf Version and Debug Info Version.

Comment (damyanp) [reply llvm/test/CodeGen/DirectX/clean-module-flags2.ll:1]: Seems overkill to me - and by this precedent we'd need to add new combinations of flags is new ones ever show up. But probably doesn't hurt.

**Review by damyanp — APPROVED** [review]

File: llvm/test/CodeGen/DirectX/clean-module-flags.ll (hunk 2 above)
Comment (llvm-beanz) [reply llvm/test/CodeGen/DirectX/clean-module-flags.ll:4]: This check doesn't seem correct. The code change here doesn't remove the flag it changes the behavior value. We should check for the updated behavior not check that the old behavior doesn't exist.

File: llvm/test/CodeGen/DirectX/clean-module-flags2.ll (hunk 1 above)
Comment (llvm-beanz) [reply llvm/test/CodeGen/DirectX/clean-module-flags2.ll:1]: Can we give this new test a more descriptive name? This is actually testing updating the flag behavior field, so how about "update-module-flag-behavior.ll"?

Comment (python3kgae) [reply llvm/test/CodeGen/DirectX/clean-module-flags2.ll]: Updated to legalize module flag.

**Comment by bob80905** [comment]
Could you explain why the flag behavior is set to warning? If I understand, you have 6 options to choose from, so why pick that one?

**Comment by bob80905** [comment]
> Could you explain why the flag behavior is set to warning? If I understand, you have 6 options to choose from, so why pick that one?

I read the issue, I think I get it now.

**Review by bob80905 — APPROVED** [review]

**Pull request merged**